lib/min_heap.rb

```diff
@@ -44,28 +55,55 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1)
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store.length == 0
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(logn)
+  # Space complexity: O(1)
   def heap_up(index)
-    
+    if index == 0
+      return
+    end
+    parent_index = ((index -1 )/2).floor()
+    if @store[index].key < @store[parent_index].key
+      swap(index, parent_index)
+      heap_up(parent_index)
+    end
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    if index >= @store.length
+      return
+    end
+
+    left_child_index = (2 * index + 1)
+    right_child_index = (2 * index + 2)
+
+    if left_child_index >= @store.length
+      return
+    elsif right_child_index >= @store.length
+      return
+    end
```

Comment:
What if the left child index is less than the length and the value at the left child index is less than the current parent!

```suggestion
    elsif right_child_index >= @store.length && @store[left_child_index].key >= @store[index].key
      return
    end
```